AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/commands/teleport/Back.java

```diff
@@ -40,14 +40,45 @@ public boolean onCommand(@NotNull CommandSender sender, @NotNull Command command
                         return true;
                     }
                     double originalY = loc.getY();
-                    while (!airMaterials.contains(loc.getBlock().getType().name())) {
+                    double originalX = loc.getX();
+                    double originalZ = loc.getZ();
+                    int radius = NewConfig.get().BACK_SEARCH_RADIUS.get();
+                    ArrayList<Location> possiblelocs = new ArrayList<>();
+
+                    /*
+                     * I note, that this loop may has up to radius^3 Calculations. But it fixes the issue, that users
+                     * complain, that they are ported at a different location.
+                     * Furthermore the default search radius of 5 equals 125 calculations which is acceptable
+                     */
+                    Location t = new Location(loc.getWorld(), originalX,originalY,originalZ);
+                    for(int dx = -radius; dx<= radius; dx++){
+                        t.setX(originalX-dx);
+                        for(int dz=-radius; dz<= radius; dz++){
+                            t.setZ(originalZ-dz);
+                            for(int dy= -radius; dy <= radius ; dy++){
+                                t.setX(originalY-dy);
+                                if(airMaterials.contains(t.getBlock().getType().name())){
+                                    possiblelocs.add(new Location(loc.getWorld(), originalX-dx, originalY-dy, originalZ-dz));
+                                }
+                            }
+                        }
+                    }
+                    while(possiblelocs.size()> 1){
+                        if(loc.distance(possiblelocs.get(1)) > loc.distance(possiblelocs.get(0))) possiblelocs.remove(1);
+                        else possiblelocs.remove(0);
+                    }
+                    if(possiblelocs.size() == 1) loc = possiblelocs.get(0);
+                    int lavablocks = 0;
+                    while (!airMaterials.contains(loc.getBlock().getType().name()) && possiblelocs.isEmpty()) {
                         // If we go beyond max height, stop and reset the Y value
-                        if (loc.getY() > loc.getWorld().getMaxHeight()) {
+                        if(loc.getBlock().getType().name().equalsIgnoreCase("Lava")) ++lavablocks;
+                        if (loc.getY() > loc.getWorld().getMaxHeight() || lavablocks > 5 ) {
```

Comment:
Yeah that's a good question. The idea is, that if there are more than 5 lava blocks the chance is big, that the player will spawn above a lava lake. Which I believe is not a good choice...

Reply:
That's fair - do you think the same should apply for normal water as well or would that be a lesser concern that should only be addressed should complaints surface?

Reply:
It's a good question, I think no, because if you got teleported over a water lake you do not die if you just get dropped in the lake, whilst you do if you get dropped in lava...